apollo-router/src/batching.rs

```diff
@@ -5,129 +5,247 @@
 //! `Batch`, there are a series of utility functions for efficiently converting
 //! graphql Requests to/from batch representation in a variety of formats: JSON, bytes
 
-use std::collections::HashMap;
 use std::fmt;
-use std::sync::Arc;
 
 use bytes::BufMut;
 use bytes::BytesMut;
 use hyper::Body;
-use parking_lot::Mutex;
+use tokio::sync::mpsc;
 use tokio::sync::oneshot;
+use tokio::task::JoinHandle;
 use tower::BoxError;
 
+use crate::error::FetchError;
 use crate::graphql;
 use crate::services::SubgraphRequest;
 use crate::services::SubgraphResponse;
 use crate::Context;
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
 pub(crate) struct BatchQuery {
+    /// The index of this query relative to the entire batch
     index: usize,
-    // Shared Batch
-    shared: Arc<Mutex<Batch>>,
+
+    /// A channel sender for sending updates to the entire batch
+    sender: mpsc::Sender<BatchHandlerMessage>,
```

Comment:
If my idea about dropping this once the `BatchQuery` is done talking to the entire batch is a good one, this will need to be an `Option`.